### PR TITLE
Add new sub-package dnf-utils

### PR DESCRIPTION
**Pull request by Conan-Kudo**
Revised version of #197 that moves the main binary to `/usr/libexec/dnf-utils`.

Requires: #199
Requires fedora release of yum-utils-1.1.31-513 with deprecated binaries - PR rpm-software-management/yum-utils#16
Obsoletes: #197

### CODE REVIEW

File: libexec/dnf-utils.in [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+           'repo-graph': ['repograph'],
+           'repoclosure': ['repoclosure'],
+           'repomanage': ['repomanage'],
+           'repoquery': ['repoquery'],
+           'reposync': ['reposync'],
+           'repotrack': ['download'],
+           'yum-builddep': ['builddep'],
+           'yum-config-manager': ['config-manager'],
+           'yum-debug-dump': ['debug-dump'],
+           'yum-debug-restore': ['debug-dump'],
```
Comment (ignatenkobrain) [inline libexec/dnf-utils.in]: `debug-restore`

File: libexec/dnf-utils.in [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+# Red Hat, Inc.
+#
+
+from __future__ import unicode_literals
+import sys
+import os.path
+
+MAPPING = {'debuginfo-install': ['debuginfo-install'],
+           'find-repos-of-install': ['list', 'installed'],
+           'package-cleanup': [],
```
Comment (ignatenkobrain) [inline libexec/dnf-utils.in]: meh... I don't think that having empty list is good idea

File: libexec/dnf-utils.in [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+           'repo-graph': ['repograph'],
+           'repoclosure': ['repoclosure'],
+           'repomanage': ['repomanage'],
+           'repoquery': ['repoquery'],
+           'reposync': ['reposync'],
+           'repotrack': ['download'],
```
Comment (ignatenkobrain) [inline libexec/dnf-utils.in]: is it 1-1 replacement?

File: libexec/dnf-utils.in (hunk 2 above)
Comment (ignatenkobrain) [inline libexec/dnf-utils.in]: ah, see it now

File: libexec/dnf-utils.in [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+from dnf.cli import main
+command = os.path.basename(sys.argv[0])
+args = sys.argv[1:]
+
+if command == 'package-cleanup':
+    if '--dupes' in args:
+        args[args.ingex('--dupes')] = '--duplicates'
+        MAPPING[command] = ['repoquery']
+    elif '--leaves' in args:
```
Comment (ignatenkobrain) [inline libexec/dnf-utils.in]: what if user passes all of those arguments? probably we should use proper argparse here?

File: dnf-plugins-core.spec [hunk 5]
```diff
@@ -308,6 +325,28 @@ pushd build-py3
...
 %find_lang %{name}
+%if %{with python3}
+mv %{buildroot}%{_libexecdir}/dnf-utils-3 %{buildroot}%{_libexecdir}/dnf-utils
+%else
+mv %{buildroot}%{_libexecdir}/dnf-utils-2 %{buildroot}%{_libexecdir}/dnf-utils
+%endif
+rm -vf %{buildroot}%{_libexecdir}/dnf-utils-*
+
+mkdir -p %{buildroot}%{_bindir}
+ln -sf %{_libexecdir}/dnf-utils %{buildroot}%{_bindir}/debuginfo-install
```
Comment (ignatenkobrain) [inline dnf-plugins-core.spec]: can we just do `for` loop here instead?

Comment (Conan-Kudo) [inline dnf-plugins-core.spec]: I generally don't like using shell loops as they're finicky...

File: libexec/dnf-utils.in (hunk 3 above)
Comment (Conan-Kudo) [inline libexec/dnf-utils.in]: `repotrack` is essentially `dnf download --resolve`, I think.

File: libexec/dnf-utils.in (hunk 4 above)
Comment (Conan-Kudo) [inline libexec/dnf-utils.in]: Yeah, probably, but what are the rules for `package-cleanup`?

File: libexec/dnf-utils.in [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    elif '--cleandupes' in args:
+        args[args.ingex('--cleandupes')] = '--duplicates'
+        MAPPING[command] = ['remove']
+    elif '--oldkernels' in args:
+        args[args.ingex('--oldkernels')] = '--oldinstallonly'
+        MAPPING[command] = ['remove']
+    else:
+        sys.stderr.write('package-cleanup has to be executed with one of the options: --dupes, '
+                         '--leaves, --orphans, --oldkernels, --problems, --cleandupes, '
+                         'or --oldkernels')
```
Comment (j-mracek) [inline libexec/dnf-utils.in]: string should end with '\n'

File: libexec/dnf-utils.in (hunk 3 above)
Comment (j-mracek) [inline libexec/dnf-utils.in]: @Conan-Kudo Unfortunately not, it is like: ``dnf download <package> $(dnf repoquery --requires <package> --resolve) `` but how to implement it?

**Comment by j-mracek** [comment]
We should also consider that in rawhide there are packages that depend on yum-utils. I know that dnf-utils is optional package, but it breaks things mayby after deprecation of yum-utils binaries.
```
[hrom@localhost ~]$ sudo dnf repoquery --whatrequires yum-utils
Last metadata expiration check: 0:00:42 ago on Fri May 12 11:28:39 2017 CEST.
copr-backend-0:1.99-1.fc27.noarch
diskimage-builder-0:1.26.1-2.fc26.noarch
fedora-gooey-karma-0:0.1-8.fc26.noarch
lpf-0:0.2-4.f1f5dd9.fc26.noarch
mock-0:1.4.1-1.fc27.noarch
pungi-0:4.1.15-1.fc27.noarch
retrace-server-0:1.17.0-1.fc27.noarch
salt-0:2016.11.5-2.fc27.noarch
yum-plugin-aliases-0:1.1.31-512.fc26.noarch
```

File: dnf-plugins-core.spec [hunk 7]
```diff
@@ -121,6 +121,23 @@ copr, debug, debuginfo-install, download, needs-restarting, repoclosure, repogra
 reposync commands. Additionally provides generate_completion_cache passive plugin.
 %endif
 
+%package -n dnf-utils
+Conflicts:      yum-utils < 1.1.32
```
Comment (j-mracek) [inline dnf-plugins-core.spec]: Should be changed to:
```yum-utils < 1.1.31-513```

File: dnf-plugins-core.spec [hunk 8]
```diff
@@ -121,6 +121,23 @@ copr, debug, debuginfo-install, download, needs-restarting, repoclosure, repogra
...
+Requires:       python3-dnf < %{dnf_not_compatible}
+%else
+Requires:       python2-dnf >= %{dnf_lowest_compatible}
+Requires:       python2-dnf < %{dnf_not_compatible}
+%endif
+Summary:        Yum-utils CLI compatibility layer
+
+%description -n dnf-utils
+As a Yum-utils CLI compatibility layer, supplies in /usr/bin/ builddep, repoquery, repoclosure,
+download, debuginfo-install, and config-manager and redirecting to DNF.
```
Comment (j-mracek) [inline dnf-plugins-core.spec]: It doesn't mention all redirections.

**Comment by Conan-Kudo** [comment]
@j-mracek @ignatenkobrain I believe I've improved it as much I reasonably should for this particular PR. My local testing of this on Mageia (where I'm already shipping this) has been very positive so far.

I would like to have this merged in ASAP and then we can look at some of the more drastic improvments (like changing the shim to use argparse, which I'm not sure we can do..).

**Comment by Conan-Kudo** [comment]
@j-mracek As for things that depend on yum-utils, they need to be patched to use dnf anyway, or the Requires that they have should be for the specific binaries they need.

At least I already fixed Mock, so once this is merged, we can work with @xsuchy to backport into Fedora.

**Comment by dmach** [comment]
@j-mracek as the dnf-utils package is optional and a user has to install it by hand to get to any trouble with packages depending on yum-utils, I'd merge this pull request.

**Comment by j-mracek** [comment]
Then thanks all for cooperation.

**Comment by j-mracek** [comment]
@dnf-bot r+

**Pull request closed without merging**